From 28307ff1aaae1d595a78f12eb69a8779207148e9 Mon Sep 17 00:00:00 2001
From: Alejandro Abdelnur <tucu@apache.org>
Date: Wed, 6 Mar 2013 00:50:32 +0000
Subject: [PATCH 1291/1357] MR1: MAPREDUCE-5028. Maps fail when io.sort.mb is set to high value. (kkambatl via tucu)

git-svn-id: https://svn.apache.org/repos/asf/hadoop/common/branches/branch-1@1453107 13f79535-47bb-0310-9956-ffa450edef68
(cherry picked from commit 2ce6dc93376cb7bd54d7adafb49127151bcb686f)

Conflicts:
	src/mapred/org/apache/hadoop/mapreduce/ReduceContext.java

Reason: Integer overflow in combiner
Ref: CDH-10562
Author: Karthik Kambatla
---
 .../src/mapred/org/apache/hadoop/mapred/Task.java  |    6 ++++--
 1 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/hadoop-mapreduce1-project/src/mapred/org/apache/hadoop/mapred/Task.java b/hadoop-mapreduce1-project/src/mapred/org/apache/hadoop/mapred/Task.java
index 0fc6a22..921f7d2 100644
--- a/hadoop-mapreduce1-project/src/mapred/org/apache/hadoop/mapred/Task.java
+++ b/hadoop-mapreduce1-project/src/mapred/org/apache/hadoop/mapred/Task.java
@@ -1222,7 +1222,8 @@ abstract public class Task implements Writable, Configurable {
       more = in.next();
       if (more) {
         DataInputBuffer nextKeyBytes = in.getKey();
-        keyIn.reset(nextKeyBytes.getData(), nextKeyBytes.getPosition(), nextKeyBytes.getLength());
+        keyIn.reset(nextKeyBytes.getData(), nextKeyBytes.getPosition(),
+            nextKeyBytes.getLength() - nextKeyBytes.getPosition());
         nextKey = keyDeserializer.deserialize(nextKey);
         hasNext = key != null && (comparator.compare(key, nextKey) == 0);
       } else {
@@ -1236,7 +1237,8 @@ abstract public class Task implements Writable, Configurable {
      */
     private void readNextValue() throws IOException {
       DataInputBuffer nextValueBytes = in.getValue();
-      valueIn.reset(nextValueBytes.getData(), nextValueBytes.getPosition(), nextValueBytes.getLength());
+      valueIn.reset(nextValueBytes.getData(), nextValueBytes.getPosition(),
+          nextValueBytes.getLength() - nextValueBytes.getPosition());
       value = valDeserializer.deserialize(value);
     }
   }
-- 
1.7.0.4

